nmt.go

```diff
@@ -142,71 +149,119 @@ func (n *NamespacedMerkleTree) ProveRange(start, end int) (Proof, error) {
 
 // ProveNamespace returns a range proof for the given NamespaceID.
 //
-// In case the underlying tree contains leaves with the given namespace
-// their start and end index will be returned together with a range proof and
-// the found leaves. In that case the returned leafHash will be nil.
+// case 1) If the namespace nID is out of the range of the tree's min and max namespace
+// i.e., (nID < n.minNID) or (n.maxNID < nID)
+//  we do not generate any range proof, instead we return an empty Proof with the range (0,0) i.e.,
+// Proof.start = 0 and Proof.end = 0
+// to indicate that this namespace is not contained in the tree.
 //
-// If the tree does not have any entries with the given Namespace ID,
-// but the namespace is within the range of the tree's min and max namespace,
-// this will be proven by returning the (namespaced or rather flagged)
-// hash of the leaf that is in the range instead of the namespace.
+// case 2) If the namespace nID is within the range of the tree's min and max namespace
+// i.e., n.minNID<= n.ID <=n.maxNID
+// and the tree does not have any entries with the given Namespace ID nID,
+// this will be proven by returning the inclusion/range Proof of the (namespaced or rather flagged)
+// hash of the leaf of the tree with the largest namespace ID that is smaller than nID.
+// if there are multiple of such leaves, then the proof is done w.r.t. to the one with the highest index
+// the leafHash field of the returned Proof will contain the namespaced hash of such leaf.
 //
-// In the case (nID < minNID) or (maxNID < nID) we do not
-// generate any proof and we return an empty range (0,0) to
-// indicate that this namespace is not contained in the tree.
+// case 3) In case the underlying tree contains leaves with the given namespace
+// their start and end index will be returned together with a range proof and
+// the found leaves. In that case the leafHash field of the returned Proof will be nil.
 func (n *NamespacedMerkleTree) ProveNamespace(nID namespace.ID) (Proof, error) {
 	isMaxNsIgnored := n.treeHasher.IsMaxNamespaceIDIgnored()
-	// In the cases (nID < minNID) or (maxNID < nID),
-	// return empty range and no proof:
-	if nID.Less(n.minNID) || n.maxNID.Less(nID) {
+	// case 1)
+	// In the cases (n.nID < minNID) or (n.maxNID < nID),
+	// return empty range and no proof
+	if nID.Less(n.minNID) || n.maxNID.Less(nID) { // TODO [Me] we could move this entire if block inside the `foundInRange` function
+		// TODO [Me] Shouldn't we instead return the first or the last node in the tree as the exclusion proof?
+		// TODO [Me] although I think this current logic is based on the premise that the root of the tree is trusted
```

Comment:
That is a good question. I thought the root (which commits/includes the min/max nids of the tree) is proof enough that the tree does not contain the passed-in `nID`. Is there a way to construct a tree/root such that the root is not sufficient proof (that the namespace is not included)? I do not think so but maybe I am missing sth. These are exactly the edge-cases that might require further thoughts.

Reply:
If all the querying nodes have access to the correct NMT roots, then there should be no problem hence no exclusion proof is required (I will still think about it to see if any corner cases may happen). Thanks for your answer!